Fixes #37717 - Update evr extension ownership to foreman

sjha4 opened the pull request:
Update ownership of evr extension to owner foreman to allow migrations to operate on it, specifically disable it.

sjha4 committed:
Fixes #37717 - Update evr extension ownership to foreman

ianballou commented:
@ekohl mentioned before that the `pg_shdepend` table might also need updating, but at least functionally it doesn't seem necessary since the migration passes with @sjha4 's changes alone. The bits that would get updated in `pg_shdepend` may get dropped with the evr extension.

ianballou commented:
I confirmed that dropping the extension also drops the related entry from `pg_shdepend`.

sjha4 commented:
I tested the installer with the chages here although I had to put the new code in a separate pre-hook file to be run..Suggestions on where to put this code if not this file?

ekohl commented on hooks/boot/06-postgresql-upgrade-extensions.rb:
```diff
@@ -43,6 +43,11 @@ def postgresql_upgrade(new_version)
 
     execute_as!('postgres', 'vacuumdb --all --analyze-in-stages', false, true)
 
+    if new_version.to_i == 13
```
We already had users upgrade their installs on Foreman 3.11 so I don't think this is the right place. Even if we cherry pick this to 3.11, we'll still need a solution for existing Katello users.

I think you need a separate hook. You should check if the `postgresql-evr` package is installed before you execute it.

sjha4 commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,5 @@
+result = execute!("runuser -l postgres -c \"psql -d foreman -c \\\"SELECT 1 from pg_extension WHERE extname='evr';\\\"\"", false, true)
+unless (result.include?('(0 rows)'))
+  execute!("runuser -l postgres -c \"psql -d foreman -c \\\"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='foreman') WHERE extname='evr';\\\"\"", false, true)
+end
```
@ekohl : Updated to check if extension evr exists in db. This check seems more relevant to updating ownership. The update ownership query itself is idempotent and doesn't fail if evr extension isn't enabled so we don't need any checks before we run it but we can keep the checks if preferred.

Let me know if this works.

sjha4 commented on hooks/boot/01-kafo-hook-extensions.rb:
```diff
@@ -113,6 +113,7 @@ def execute!(command, do_say = true, do_log = true)
       exit 1
     end
+    stdout_stderr
   end
```
Making execute! return the output here cause otherwise it's nil and calling function doesn't have access to the output.

evgeni commented:
What about remote db users? Will those need to perform manual steps (we usually don't have enough permissions to change ownership of objects there) before the upgrade?

sjha4 commented:
> What about remote db users? Will those need to perform manual steps (we usually don't have enough permissions to change ownership of objects there) before the upgrade?

Ya..They'll need to run the update ownership query manually as part of their upgrades. Will need to be documented.

ianballou commented:
@evgeni docs update will go in https://github.com/theforeman/foreman-documentation/pull/3167 unless someone gets to it first.

ianballou commented:
Prepping a system for a test.

ianballou commented:
Found an issue -- the postgres command cannot run because the postgres service is down during the installer run. We'll need to turn postgres on first.

ianballou commented:
Aside from the postgres issue it's working.

sjha4 commented:
> Aside from the postgres issue it's working.

Ok..So couple of changes..
1. Added a check to run only on local postgres.
2. Added code to check status of postgresql service, start it if stopped and stop it after the hook.

Should work better now.

ianballou commented:
Tested again by upgrading Katello 4.12 to nightly via 4.13. Worked well.

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
Please respect noop mode, which you can check with `app_value(:noop)`

sjha4 replied:
Updated code to respect no-op and address other review comments..

sjha4 replied:
Question: What are these numbers before the filenames? I picked 35 cause 34- existed. 🙊

sjha4 replied:
Looks like foreman release versions..I can update the filename to 311- to follow convention if that is what it is..

ekohl replied:
They're for order. All hooks get sorted and that determines when they're executed. It allows you to ensure your hook is executed after the PostgreSQL migration itself

ekohl replied:
I'd prefer to implement the noop mode slightly different. In our other hooks we log debug statements what would be done. Otherwise you can simply reduce it to `if !app_value(:noop) &&local_postgresql? && execute("rpm -q postgresql-evr", false, false)`.

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,17 @@
+if local_postgresql?
+  response, _status = execute_command("sudo systemctl is-active postgresql", false, true)
```
The installer already runs as root, so no need for sudo
```suggestion
  response, _status = execute_command("systemctl is-active postgresql", false, true)
```

sjha4 replied:
Updated..

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,17 @@
+  end
+
+  logger.notice("Updating ownership of the evr extension if it is enabled")
+  result = execute!("runuser -l postgres -c \"psql -d foreman -c \\\"SELECT 1 from pg_extension WHERE extname='evr';\\\"\"", false, true)
```
Is this needed? I'd expect that the `UPDATE` line is simply a noop if it's not installed.

sjha4 replied:
Removed the checks..The query is a no-op if evr extension isn't found..

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,17 @@
+if local_postgresql?
```
I'd also limit it to when `postgresql-evr` is installed
```suggestion
if local_postgresql? && execute("rpm -q postgresql-evr", false, false)
```

We already have that code in 2 other places so perhaps it's also time for a common wrapper method like `package_installed?` `oauth_aio_gem_installed?` already has most of that logic anyway

sjha4 replied:
Updated with suggested code..Didn't mess with wrapper right now..

sjha4 commented:
Not sure if test failure is related.. 😕

ehelms commented:
> Not sure if test failure is related.. 😕

Looks like minitar 1.0.0 was released today and we made need to pin it for the time being -- https://rubygems.org/gems/minitar/versions/1.0.0

sjha4 commented:
> > Not sure if test failure is related.. 😕
> 
> Looks like minitar 1.0.0 was released today and we made need to pin it for the time being -- https://rubygems.org/gems/minitar/versions/1.0.0

Ya..That was it..

https://github.com/theforeman/foreman-installer/pull/956

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,24 @@
+# In Katello 4.14, the 'evr' extension is removed from PostgreSQL and integrated into the Katello database via a migration.
+# This hook ensures the 'evr' extension's ownership is transferred to the 'foreman' user so migrations can act on it.
+
+if local_postgresql? && execute("rpm -q postgresql-evr", false, false)
```
I think we should also have an `else` for `local_postgresql?` where we check if the extension exists with an incorrect owner. We have all the parameters to instruct `psql` to connect to the right location. You can't fix it, but you can halt the installer to prevent a half-upgraded system. Though that should probably be done in a `pre_commit` hook.

https://github.com/theforeman/puppet-foreman/blob/5f38479c870ef03a4a1565007be68b61e8950d53/manifests/init.pp#L73-L85 are the various answers. foreman_maintain has code that already does something similar, but based on `database.yml`:
https://github.com/theforeman/foreman_maintain/blob/6cc814b8c669bd02a15452fd488a0a6221ed1fd8/lib/foreman_maintain/concerns/base_database.rb#L148-L156

sjha4 replied:
Circling back to this, I will look at a pre-commit hook for non-local DBs to get this moving..

ianballou replied:
I can add the pre_commit hook in a separate PR to keep this one more focused.

sjha4 replied:
@ekohl : Ian's got a PR out : https://github.com/theforeman/foreman-installer/pull/984/ to add the pre-commit hook to check ownership on remote dbs. The local DBs would be handled with this PR..

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,24 @@
+  # Update the ownership of the evr extension
+  unless app_value(:noop)
+    logger.debug("Updating ownership of the evr extension if it is enabled")
+    execute!("runuser -l postgres -c \"psql -d foreman -c \\\"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='foreman') WHERE extname='evr';\\\"\"", false, true)
```
First of all, it can be simplified to use `execute_as!`. And I think you should look at the answers as well. Then the final version is:
```suggestion
    database = param_value('foreman', 'db_database')
    username = param_value('foreman', 'db_username')
    sql = "UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='#{username}') WHERE extname='evr'"
    execute_as!('postgres', "psql -d #{database} -c \"#{sql}\"", false, true)
```

But escaping will be tricky, so I'd enhance `execute_as!` to pass it as an array:
```ruby
def execute_as!(user, command, do_say = true, do_log = true)
  runuser_command = ['runuser', '-l', user, '-c', command] 
  execute!(runuser_command, do_say, do_log)
end
```

Then I'd expect it to work if users chose a different DB and/or username as well.

ekohl replied:
I just realized that `Kafo::PuppetCommand.format_command` probably doesn't deal with commands as arrays.

sjha4 replied:
Updated to do this:
```
  database = param_value('foreman', 'db_database') || 'foreman'
  username = param_value('foreman', 'db_username') || 'foreman'
  sql = "runuser -l postgres -c \"psql -d '#{database}' -c \\\"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='#{username}') WHERE extname='evr';\\\"\""
  logger.debug("Executing: #{sql}")
  execute!(sql, false, true) unless app_value(:noop)
```

Still using execute! as escaping was being really painful.

ekohl replied:
I opened https://github.com/theforeman/kafo/pull/374 for that support.

sjha4 replied:
Unresolving this and I am going to push an update to use this cause I was running into some escaping issues on the sql command execution.

ianballou commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,26 @@
+# In Katello 4.14, the 'evr' extension is removed from PostgreSQL and integrated into the Katello database via a migration.
```
```suggestion
# In Katello 4.15, the 'evr' extension is removed from PostgreSQL and integrated into the Katello database via a migration.
```

sjha4 commented:
@ekohl : I am not sure I remember if we were supposed to change anything here following the kafo updates..

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,26 @@
+# This hook ensures the 'evr' extension's ownership is transferred to the 'foreman' user so migrations can act on it.
+
+if local_postgresql? && execute("rpm -q postgresql-evr", false, false)
+  is_postgresql_active = execute_command("systemctl is-active postgresql", false, true)&.first&.strip == "active"
```
Isn't it better to check the exit code is 0 instead of parsing the output? I have no idea if systemctl has localization, but that's one example where it could break.

sjha4 replied:
Ack..Updated to check exit status code

ekohl reviewed (commented):
I'm not sure about the noop mode implementation here. I'd be tempted to change the log messages to `Would do X` instead of Doing x`.

ehelms commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,26 @@
+  logger.debug("Updating ownership of the evr extension if it is enabled")
+  database = param_value('foreman', 'db_database') || 'foreman'
+  username = param_value('foreman', 'db_username') || 'foreman'
+  sql = "psql -d '#{database}' -c \\\"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='#{username}') WHERE extname='evr';\\\""
```
Can you re-factor the way the SQL is handled to re-use the functions added via https://github.com/theforeman/foreman-installer/pull/984/files to ensure no passwords are leaked?

ekohl replied:
This one is a bit different in that we run it as `postgres` (because we assume that's a superuser), relying on ident auth so no actual passwords are used. Since we don't consider the DB name and username to be that secret, it probably is ok.

ehelms reviewed (approved):
LGTM -- I'll let @ekohl have the final say.

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,26 @@
+# In Katello 4.16, the 'evr' extension is removed from PostgreSQL and integrated into the Katello database via a migration.
+# This hook ensures the 'evr' extension's ownership is transferred to the 'foreman' user so migrations can act on it.
+
+if local_postgresql? && execute("rpm -q postgresql-evr", false, false)
```
Probably better to specifically check that Foreman is local instead of there being any local DB:
```suggestion
if local_foreman_db? && execute("rpm -q postgresql-evr", false, false)
```

Though perhaps you can also include `devel_scenario?` in there if you want to accommodate for the katello-devel boxes.

sjha4 replied:
Updated to 
```
if (local_foreman_db? || devel_scenario?) && execute("rpm -q postgresql-evr", false, false)
```

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,26 @@
+# This hook ensures the 'evr' extension's ownership is transferred to the 'foreman' user so migrations can act on it.
+
+if local_postgresql? && execute("rpm -q postgresql-evr", false, false)
+  is_postgresql_active = execute_command("systemctl is-active postgresql", false, true)&.last&.exitstatus == 0
```
I'm debating simplification and wonder what you think. And I'm only talking about the `op` "mode", not about `noop`.

You can always call `start_services()` even if it was already running, so the check really is only useful for stopping. But do we need to stop the service? The installer will later start it anyway. Perhaps just leave it running?

sjha4 replied:
Hmm..The idea was to leave the service in the state this hook finds it in..I don't feel strongly about leaving it running after this hook cause it would get stopped by anything after this that needs it stopped. I'll push to simplify the thing..

ekohl reviewed (commented):
Style wise I'm not that happy with the many conditionals and prefer the more verbose:
```ruby
if app_value(:noop)
  logger.debug('Would do X')
else
  logger.debug('Doing X')
  x()
end
```

My reasoning is that it's very easy to visually see the control flow.

Additionally, we already have debug statements in various methods, like `start_services` and `stop_services`. So you can simplify it to:

```ruby
if app_value(:noop)
  logger.debug('Would stop service postgresql')
else
  stop_services(['postgresql'])
end
```

I'm wondering what you think about this.

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,21 @@
+
+if (local_foreman_db? || devel_scenario?) && execute("rpm -q postgresql-evr", false, false)
+  if app_value(:noop)
+    logger.debug("Would start postgresql service") if app_value(:noop)
```
```suggestion
    logger.debug("Would start postgresql service")
```

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,21 @@
+    logger.debug("Would execute: #{sql} to update ownership of the evr extension")
+  else
+    logger.debug("Updating ownership of the evr extension")
+    execute_as!('postgres', sql, false, true)
```
This has me thinking: how could this fail and how would a user resolve it?

One I can think of is that the user has `postgresql-evr` installed but the DB doesn't actually exist. I'm wondering if we need to catch that. Would that affect backup restores?

Any others that you can think of?

ekohl replied:
Thinking out loud further: there's `--reset-data` which could interact in interesting ways

sjha4 replied:
Tested --reset-data and it was failing..I added a check to confirm if local foreman database exists before proceeding.

For backup-restore, I'd assume the supported workflow would be databases at the same version of foreman where the migrations in [katello PR](https://github.com/Katello/katello/pull/11159) have run.

ekohl reviewed (commented):
Thanks for updating. I think this is much easier to reason about and see where it could fail.

ekohl commented on hooks/boot/01-kafo-hook-extensions.rb:
```diff
@@ -164,6 +164,11 @@ def execute_as!(user, command, do_say = true, do_log = true, extra_env = {})
     execute!(runuser_command, do_say, do_log, extra_env)
   end
 
+  def execute_as_formatted!(user, command, do_say = true, do_log = true, extra_env = {})
```
I like using https://github.com/theforeman/kafo/commit/c463e3f38f5477a94460f1b7dbd8d8eedd87801d to greatly reduce the amount of escaping needed, but have some questions about the implementation.

It's called `execute_as_formatted!` and `!` implies it raises an exception if it fails (copied from Rails, see https://stackoverflow.com/questions/612189/why-are-exclamation-marks-used-in-ruby-methods for more discussion on the convention). `execute_command` doesn't, so that's a mismatch.

On the name itself, I think formatted could be confusing because I don't see any formatting. Naming is hard, but any thoughts on `execute_as_user_verbatim`?

sjha4 replied:
How does `execute_preformatted_as` sound?

ekohl commented on hooks/boot/01-kafo-hook-extensions.rb:
```diff
@@ -164,6 +164,11 @@ def execute_as!(user, command, do_say = true, do_log = true, extra_env = {})
 
+  def execute_as_formatted!(user, command, do_say = true, do_log = true, extra_env = {})
+    runuser_command = ['runuser', '-l', user, '-c', command]
+    execute_command(runuser_command, do_say, do_log, extra_env)
```
I wonder if `do_say` and `do_log` can deal with an array properly.

sjha4 replied:
```
2024-12-03 12:38:14 [DEBUG ] [pre] Updating ownership of the evr extension
2024-12-03 12:38:14 [DEBUG ] [root] Executing: ["runuser", "-l", "postgres", "-c", "psql -d foreman -c \"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='foreman') WHERE extname='evr'\""]
2024-12-03 12:38:14 [DEBUG ] [root] UPDATE 1
```

This is how it's logged..

sjha4 replied:
I'll update logging to join the array and log so it's displayed properly.

ekohl commented on hooks/boot/01-kafo-hook-extensions.rb:
```diff
@@ -164,13 +170,19 @@ def execute_as!(user, command, do_say = true, do_log = true, extra_env = {})
 
   def execute_command(command, do_say, do_log, extra_env = {})
-    log_and_say(:debug, "Executing: #{command}", do_say, do_log)
+    display_command = command.is_a?(Array) ? command.join(' ') : command
```
Should we use `shelljoin` to properly escape it?

I prefer the more explicit call:
```suggestion
    display_command = command.is_a?(Array) ? Shellwords.join(command) : command
```

But this technically also works:
```suggestion
    display_command = command.is_a?(Array) ? command.shelljoin : command
```

Be sure that `shellwords` is required somewhere

sjha4 replied:
```
2024-12-03 13:47:28 [DEBUG ] [pre] Starting services postgresql
2024-12-03 13:47:28 [DEBUG ] [root] Executing: runuser -l postgres -c psql\ -t\ -c\ \"SELECT\ 1\ FROM\ pg_database\ WHERE\ datname\ \=\ \'foreman\'\;\"
2024-12-03 13:47:28 [DEBUG ] [root]         1
2024-12-03 13:47:28 [DEBUG ] [root] 
2024-12-03 13:47:28 [DEBUG ] [pre] Updating ownership of the evr extension
2024-12-03 13:47:28 [DEBUG ] [root] Executing: runuser -l postgres -c psql\ -d\ foreman\ -c\ \"UPDATE\ pg_extension\ SET\ extowner\ \=\ \(SELECT\ oid\ FROM\ pg_authid\ WHERE\ rolname\=\'foreman\'\)\ WHERE\ extname\=\'evr\'\"
2024-12-03 13:47:28 [DEBUG ] [root] UPDATE 1
2024-12-03 13:47:28 [DEBUG ] [pre] Hook /usr/share/foreman-installer/hooks/pre/35-change-evr-extension-ownership.rb returned nil
```
Confirmed that both the shell formatted strings can be run directly on the terminal although the number of `\` is making it a little hard to read. 🤷🏼

ekohl commented on hooks/boot/01-kafo-hook-extensions.rb:
```diff
@@ -85,6 +85,12 @@ def pulpcore_enabled?
     module_enabled?('foreman_proxy_content')
   end
 
+  def local_db_exists?(database = 'foreman')
```
I don't think we need a default
```suggestion
  def local_db_exists?(database)
```

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,25 @@
+      execute_preformatted_as('postgres', "psql -d #{database} -c \"#{sql}\"", false, true)
+    end
+  else
+    logger.debug("The foreman database does not exist.")
```
Perhaps log the explicit DB name too?
```suggestion
    logger.debug("The Foreman database '#{database}' does not exist.")
```

ekohl commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,25 @@
+      logger.debug("Would execute: #{sql} to update ownership of the evr extension")
+    else
+      logger.debug("Updating ownership of the evr extension")
+      execute_preformatted_as('postgres', "psql -d #{database} -c \"#{sql}\"", false, true)
```
This will now silently fail. Should it check the status and at least log a warning?

ekohl reviewed (commented):
I like how this is shaping up now. Some minor style comments.

ekohl reviewed (approved):
@evgeni I made may suggestions so I'd appreciate if you gave it a final ACK.

evgeni commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,28 @@
+  if local_db_exists?(database)
+    sql = "UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='#{username}') WHERE extname='evr'"
+    if app_value(:noop)
+      logger.debug("Would execute: #{sql} to update ownership of the evr extension")
```
```suggestion
      logger.debug("Would execute the following SQL statement to update ownership of the evr extension: #{sql}")
```

That way it's more clear where text ends and SQL begins

sjha4 replied:
Updated to this message. One more change I hadn't noticed I had locally was updating a couple of log warnings to logger.notice because that seemed to look better in the output.

evgeni commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,28 @@
+      logger.debug("Updating ownership of the evr extension")
+      _, success = execute_preformatted_as('postgres', "psql -d #{database} -c \"#{sql}\"", false, true)
+      unless success
+        logger.notice("Failed to update ownership of the evr extension.")
```
I think failing to update should be a warning?

ekohl replied:
Warning or even error that halts execution?

evgeni replied:
Maybe. It was a warning before, so that's what caught my eye.

What happens if it is not updated? Later the removal of the extension fails?

sjha4 replied:
Ya..The evr extension is dropped in katello migration here https://github.com/Katello/katello/pull/11159/ which assumes the foreman user is the owner of the extension. We could fail here instead of later on during db:migrate.

sjha4 replied:
Updated to fail and exit:
```
fail_and_exit("Failed to update ownership of the evr extension. Please make sure the following sql succeeds before proceeding: #{sql}")
```

sjha4 replied:
@evgeni @ekohl Is this what y'all had in mind?

evgeni commented on hooks/pre/35-change-evr-extension-ownership.rb:
```diff
@@ -0,0 +1,28 @@
+      end
+    end
+  else
+    logger.notice("The Foreman database #{database} does not exist.")
```
When does this happen? When the system is not yet fully installed?

ekohl replied:
In https://github.com/theforeman/foreman-installer/pull/955#discussion_r1866425877 I suggested it. I'm not sure if we'll hit it, but I can imagine there are weird edge cases with `--reset-data`, DB restores or equivalent odd cases that I'd like to handle gracefully.

evgeni replied:
Makes sense. I don't think I care for debug vs notice here.

sjha4 commented:
I will need someone with superpowers to merge this one if this is good to go.. 
Thanks @ekohl , @ehelms , @evgeni and @ianballou  ❤️

ekohl reviewed (approved):
Thanks!
